harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,881 @@
+                            self.context.distributed._zmq_broadcast((storage_id, path))
+                            self._save(pathlib.Path(path))
+                            # Gather resources across nodes.
+                            all_resources = self.context.distributed._zmq_gather(
+                                storage.StorageManager._list_directory(path)
+                            )
+                        resources = {k: v for d in all_resources for k, v in d.items()}
+
+                        self.context._core.checkpointing._report_checkpoint(
+                            storage_id, resources, metadata
+                        )
+                        response = {"uuid": storage_id}
+                    else:
+                        storage_id, path = self.context.distributed._zmq_broadcast(None)
+                        self._save(pathlib.Path(path))
+                        # Gather resources across nodes.
+                        _ = self.context.distributed._zmq_gather(
+                            storage.StorageManager._list_directory(path)
+                        )
+                        if self.context.distributed._is_local_chief:
+                            storage_manager.post_store_path(storage_id, path)
+                        response = {}
+
+                else:
+                    raise AssertionError("Unexpected workload: {}".format(w.kind))
+
+            except det.InvalidHP as e:
+                logging.info(f"Invalid hyperparameter exception during {action}: {e}")
+                response = workload.InvalidHP()
+            response_func(response)
+
+    def get_epoch_idx(self, batch_id: int) -> int:
+        return batch_id // cast(int, self.context._epoch_len)
+
+    def _train_for_step(
+        self, step_id: int, num_batches: int, total_batches_processed: int
+    ) -> workload.Response:
+        """
+        DeepSpeed allows specifying train_batch_size, train_micro_batch_size_per_gpu, and
+        gradient_accumulation_steps. The three are related as follows:
+        train_batch_size = train_micro_batch_size * gradient_accumulation_steps.
+        Hence, if two are specified, the third can be inferred.
+
+        For pipeline parallel training, DeepSpeed will automatically interleave
+        gradient_accumulation_steps worth of micro batches in one train_batch/eval_batch call.
+
+        With the default DeepSpeed model engine (no pipeline parallel training), the backward
+        and optimizer step calls track micro batches and will automatically update model weights
+        and lr scheduler if micro batches % gradient_accumulation_steps == 0.
+
+        Comparing throughput with and without pipeline parallel is a common goal so we will
+        automatically perform gradient accumulation by default when pipeline parallelism is not
+        used.  This can be turned off by setting context.disable_auto_grad_accumulation.
```

Comment:
I don't understand your reasoning here, why does model parallel have to do with gradient accumulation?